arroyo-worker/src/connectors/filesystem/s3/source.rs

```
@@ -0,0 +1,214 @@
+                continue;
+            }
+            prev_file_found = true;
+
+            let get_res = client
+                .get_object()
+                .bucket(self.bucket.clone())
+                .key(obj_key.clone())
+                .send()
+                .await
+                .unwrap();
+            let body = get_res.body;
+
+            if self.compression == "zstd" {
+                let r = ZstdDecoder::new(BufReader::new(body.into_async_read()));
+                match self
+                    .read_file(ctx, BufReader::new(r).lines(), &obj_key, prev_lines_read)
+                    .await
+                {
+                    Some(finish_type) => return finish_type,
+                    None => {}
+                }
+            } else if self.compression == "gzip" {
+                let r = GzipDecoder::new(BufReader::new(body.into_async_read()));
+                match self
+                    .read_file(ctx, BufReader::new(r).lines(), &obj_key, prev_lines_read)
+                    .await
+                {
+                    Some(finish_type) => return finish_type,
+                    None => {}
+                }
+            } else {
+                match self
+                    .read_file(
+                        ctx,
+                        BufReader::new(body.into_async_read()).lines(),
+                        &obj_key,
+                        prev_lines_read,
+                    )
+                    .await
+                {
+                    Some(finish_type) => return finish_type,
+                    None => {}
+                }
+            }
+        }
+        info!("S3 source finished");
+        SourceFinishType::Final
+    }
+
+    async fn read_file<R: AsyncBufRead + Unpin>(
+        &mut self,
+        ctx: &mut Context<(), T>,
+        mut reader: Lines<R>,
+        obj_key: &str,
+        prev_lines_read: usize,
+    ) -> Option<SourceFinishType> {
```

This function should probably return an error, in case it can't read the file, for instance due to a network issue.